active-rfcs/0000-attr-fallthrough.md

```diff
@@ -0,0 +1,220 @@
+
+The reason for `class`, `style` and `v-on` listeners to be whitelisted is because:
+
+- They cover the most common use cases for attribute fallthrough.
+- They have close to no risk of clashing with prop names.
+- They require special merge logic instead of simple overwrites, so handling them implicitly yields more convenience value.
+
+If a functional component with optional props declaration needs to support full attribute fallthrough, it needs to declare `inheritAttrs: false`, pick the desired attrs from `props`, and merge it to the root element:
+
+``` js
+// destructure props, and use rest spread to grab unused ones as attrs.
+const Func = ({ msg, ...attrs }) => {
+  return h('div', mergeProps({ class: 'foo' }, attrs), msg)
+}
+Func.inheritAttrs = false
+```
+
+## API Deprecations
+
+- `.native` modifier for v-on will be removed.
+- `this.$listeners` will be removed.
+
+# Adoption strategy
+
+- Deprecations can be supported in the compat build:
+
+  - `.native` modifier will be a no-op and emit a warning during template compilation.
+  - `this.$listeners` can be supported with a runtime warning.
+
+- There could technically be cases where the user relies on the 2.x behavior where `inheritAttrs: false` does not affect `class` and `style`, but it should be very rare. We will have a dedicated item in the migration guide / helper to remind the developer to check for such cases.
+
+- Since functional components uses a new syntax, they will likely require manual upgrades. We should have a dedicated section for functional components in the migration guide.
+
+# Unresolved questions
+
+## Removing Unwanted Listeners
+
+With flat VNode data and the removal of `.native` modifier, all listeners are passed down to the child component as `onXXX` functions:
+
+``` html
+<foo @click="foo" @custom="bar" />
+```
+
+compiles to:
+
+``` js
+h(foo, {
+  onClick: foo,
+  onCustom: bar
+})
+```
+
+When spreading `$attrs` with `v-bind`, all parent listeners are applied to the target element as native DOM listeners. The problem is that these same listeners can also be triggered by custom events - in the above example, both a native click event and a custom one emitted by `this.$emit('click')` in the child will trigger the parent's `foo` handler. This may lead to unwanted behavior.
+
+Props do not suffer from this problem because declared props are removed from `$attrs`. Therefore we should have a similar way to "declare" emitted events from a component. Event listeners for explicitly declared events will be removed from `$attrs` and can only be triggered by custom events emitted by the component via `this.$emit`. There is currently [an open RFC for it](https://github.com/vuejs/rfcs/pull/16) by @niko278. It is complementary to this RFC but does not affect the design of this RFC, so we can leave it for consideration at a later stage, even after Vue 3 release.
```

Comment:
To support just arbitrary attributes (without event listeners) on the component root it won't be safe to simply reply on implicit fallthrough because:

1. All components have a fallthrough for `style`, `class` and `v-on` attributes
2. Given most of the components emit their custom events we'll assign unnecessary event handlers on their roots from implicit fallthrough
3. To opt-out of this behaviour we declare `inheritAttrs: false` for each component that has a custom event
4. Now to support arbitrary attributes on the root of those components we write `v-bind="filteredAttrs"` which contains attributes only

Compare that with a v2 behaviour when only `v-bind="$attrs"` is used to support arbitrary attributes on an element. And no extra work is required to work with custom events (that may have clashing names with native events).

Even if #16 does become implemented there're too many workarounds required for just custom events to be able to safely work.

----

I can hardly imagine `$attrs` containing event listeners to be as useful as `$attrs` in v2.

Consider this example:

```html
<input @input="$emit('input', $event)" v-bind="$attrs">
```

If you do this in v3 then you should get 2 `input` events at once. With `inheritAttrs: true` this should produce 3 events since we also have implicit fallthrough.

In each of those scenarios you should always filter `$attrs` and I couldn't find a good use-case where you need both attributes and listeners in one place.

Reply:
Honestly I don't understand your concerns, so I'll just answer with what I think might address it:

1. In v2 the fallthrough applies to **all attributes**. It's not just `class` and `style`. You never needed `v-bind="$attrs"` unless you use `inheritAttrs: false`. However, because `v-on` does not automatically fallthrough, users have to manually proxy the native events to custom events.

2. In v3 with the `v-on` fallthrough you should never need to explicitly proxy events again. So in most cases implicit fallthrough will just work and **you don't need to do anything**.

3. The only downside is when users listen for a custom event, a native listener is also added. This is why `emits` is suggested here: by declaring the custom events the component intends to emit, Vue can avoid binding them as native listeners.

So the only extra thing you need is the `emits` option, which can serve documentation and runtime validation purpose as well, just like `props`.

Reply:
> However, because v-on does not automatically fallthrough, users have to manually proxy the native events to custom events.

Having to add a `.native` modifier is way easier than dealing with filtering `$attrs` when authoring components. Being able to control whether an event listener should be assigned to the root or not on a component client level looks like a more reasonable approach for me, at least it does solve the issue of event listeners implicit fallthrough.

The naming could probably be reconsidered though. I suggest leaving the modifier behaviour but calling it `.root` rather than `.native` (that would better communicate that the listener would be assigned to a component's root and a fragment vnode does not fulfill this requirement).

----

The problem with this approach is that it simplifies things if you need your listeners on the root element and complicates things if you need them somewhere else. It also complicates things if you have root event listeners emitting anything but the `event` object, in that case you'll have to disable fallthrough behaviour completely and manually filter `$attrs` object.

So some components will benefit from that and others will require more work than usual. Vue 2 does not have that problem.

Maybe a balance could be achieved here?

Consider a component that has to have implicit attributes fallthrough, but at the same time emit a custom event:

```html
<input @input="$emit('input', $event.target.value)" v-bind="filteredAttrs">

<script>
  export default {
    inheritAttrs: false,
    computed: {
      filteredAttrs() {...}
    }
  }
</script>
```

The component above should benefit from attributes fallthrough in theory but it has to disable that behaviour in order to have a custom event. `filteredAttrs` should now filter all attributes except `style`, `class` and `onInput` listener.

Since #16 is not part of this RFC at the current state this is how I imagine it's supposed to work for this case when this RFC is accepted. If the new behaviour requires `emits` to deal with the issues above I think it should be a part of this RFC.

Reply:
Getting rid of `.native` is a very good thing imo

Reply:
So, the primary concern is for components that emit custom events with the same names as native events? How often does that really happen?

For most components that are wrappers of native equivalents (e.g. `<MyButton>`, `<MyInput>`), users would expect native event names to behave like native listeners. And for input components, you really should support `v-model` as the primary usage anyway, so I believe the cases where you need to emit a custom event named `input` AND the root element happens to be an `<input>` element is quite rare.

Reply:
In my codebase this happens a lot. This change would stop me from migrating to v3. Also there is not a single case of `.native` in my project, but for some reason I'll be forced to go through a migration step in order to remove it.

Reply:
@CyberAP assuming `emits` option is added, you can explicit control the events that you want to "take over". E.g. with `emits: ['input']`, `@input` will no longer fallthrough as a native listener and you can decide what to do with it. Does that resolve your problem?

Reply:
Technically a codemod can scan all instances of `$emit` calls in a component and generate the `emits` option for you.